### DISCUSSION TITLE
dynamic socket.on and middleware

### DISCUSSION THREAD

**Discussion by artstylee** [socketio/socket.io#4185] [Q&A] [answered]
hi guys! im doing a copy of discord on socket io 
and i have some questions
to listen a channel on a backend i need to do
```
socket.on('chatname', (msg) => {
    });
```
the problem is that the users dynamically add/delete channels.. 
///
another issue is .. before someone emit a message a want to check the authorization, if user are able to add/messages to this channel

the questions is :
1. how can i dynamically add/delete socket.on listeners w/o reloading the server on node.js?
2.  how can  i check if the user is authorized to add messages to the channnel?

**Comment by darrachequesne** [comment, marked as answer]
> the problem is that the users dynamically add/delete channels..

I think you can either use a common event name:

```js
socket.on("message", (channel, msg) => {
  // ...
});
```

or use a catch all listener:

```js
socket.onAny((event, ...args) => {
  // ...
});
```

Reference: https://socket.io/docs/v4/server-api/#socketonanycallback

> before someone emit a message a want to check the authorization, if user are able to add/messages to this channel

You could store the list of authorized channels when the user first connects:

```js
io.on("connection", async (socket) => {
  socket.authorizedChannels = await fetchAuthorizedChannels();

  // and then
  socket.on("message", (channel, msg) => {
    if (socket.authorizedChannels.includes(channel)) {
      // ignore the message or close the connection
    }
  });
});
```

**Reply by artstylee** [reply]
cool. thanks. i get it.
but there is another question
if i do this
`socket.authorizedChannels = await fetchAuthorizedChannels();`
it will be once on connection
but what if i want to "add to banlist" someone in real time?
i store the chat data in the database like this
```
    chatName: {type: String, required: true},
    owner: {type: String, required: true, ref: 'User'},
    mod: [{type: String, ref: 'User'}],
    banned: [{type: String, ref: 'User'}]
```
when i add someone to the banned array i want them immediately get kicked and never join the chat. could you help me with that?

**Reply by darrachequesne** [reply]
In that case, I guess you can use the `disconnectSockets()` method:

```js
io.on("connection", (socket) => {
  socket.join(theUserId);
});

// and then when you ban the user
io.in(theUserId).disconnectSockets();
```

Reference: https://socket.io/docs/v4/server-api/#serverdisconnectsocketsclose
